File: median-of-two-sorted-arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        def findMedian(a, b, a_start, a_end, b_start, b_end):
            # The target index of a + b sorted (more specifically the sub lists a + b sorted i.e. taking account of a- and b_start and a- and b_end)
            # For an even list it is the lower index of the two targets
            k = math.ceil((len(a) + len(b))/2 - 1)
            # The pivots
            # They are the middle element of an odd list or the upper middle element of an even list
            pa = a_start + math.ceil((a_end-a_start)/2)
            pb = b_start + math.ceil((b_end-b_start)/2)

            print(a[a_start:a_end+1], b[b_start:b_end+1])

            even = (len(a) + len(b)) % 2 == 0
            # If the length of a + b is even
            if even:
                # If only two elements are left
                if a_end+b_end-a_start-b_start == 0:
                    tmp = a[a_start:a_end+1] + b[b_start:b_end+1]
                    return (tmp[0]+tmp[1])/2
                # If there are only elements in either a or b
                if a_end-a_start < 0:
                    return (b[k-a_start]+b[k+1-a_start])/2
                if b_end-b_start < 0:
                    return (a[k-b_start]+a[k+1-b_start])/2
            else: # If the length of a + b is odd there are only elements in either a or b
                if a_end-a_start < 0:
                    return b[k-a_start]
                if b_end-b_start < 0:
                    return a[k-b_start]

            # For example if a[pa] < a[pb] then elements a[<=pa] are for sure among the lower_part lowest elements
            lower_part = pa+pb+1


            # What if one k is in the lower part, and the other is in the upper part?
            # This creates a problem: we can't remove any parts because we may remove one of the k's
            # We need to make sure that that does not happen
            if k == (lower_part-1):
                if a_end>pa:
                    pa += 1
                elif a_start<pa:
                    pa -= 1
                elif b_end>pb:
                    pb += 1
                elif b_start<pb:
                    pb -= 1
                lower_part = pa+pb+1

            print(a[a_start:a_end+1], b[b_start:b_end+1], k-a_start-b_start, pa-a_start, pb-b_start, lower_part-a_start-b_start)

            # Indexes < pa are in the lower part of a + b sorted, indexes > pb are in the upper part of a + b sorted
            if a[pa] <= b[pb]:
                if even:
                    # Both k's (targets) are in the lower part therefore we remove the elements that are for sure in the upper part
                    if (k+1) <= (lower_part-1):
                        return findMedian(a, b, a_start, a_end, b_start, pb-1)
                    # Both k's are in the upper part so we remove elements that are for sure in the lower part
                    else:
                        return findMedian(a, b, pa+1, a_end, b_start, b_end)
                else:
                    # If the target index is in the lower part of a + b sorted, then discard the indexes > pb
                    if k <= (lower_part-1):
                        return findMedian(a, b, a_start, a_end, b_start, pb-1)
                    # If the target index is in the upper part of a + b sorted, then discard the indexes <= pa
                    else:
                        return findMedian(a, b, pa+1, a_end, b_start, b_end)
            # Indexes > pa are in the upper part of a + b sorted, indexes < pb are in the lower part of a + b sorted
            elif a[pa] > b[pb]:
                if even:
                    if (k+1) <= (lower_part-1):
                        return findMedian(a, b, a_start, pa-1, b_start, b_end)
                    else:
                        return findMedian(a, b, a_start, a_end, pb+1, b_end)
                else:
                    if k <= (lower_part-1):
                        return findMedian(a, b, a_start, pa-1, b_start, b_end)
                    else:
                        return findMedian(a, b, a_start, a_end, pb+1 , b_end)
                    
        
        return findMedian(nums1, nums2, 0, len(nums1)-1, 0, len(nums2)-1)

File: median-of-two-sorted-arrays.cpp
#include <bits/stdc++.h>
class Solution {
public:
    double findMedianSortedArraysRec(vector<int>& nums1, vector<int>& nums2, int start1, int start2, int end1, int end2) {

        const int len = nums1.size()+nums2.size();
        const bool even = len%2==0;
        const int target = len/2; // Upper if two targets

        if (len == 0) {
            return 0.;
        }
        
        // cout <<"Arr1 " << start1 << " " << end1 << endl;
        // cout <<"Arr2 " << start2 << " " << end2 << endl;
        if (!even) {
            if (start1 > end1) {
                return nums2[target-start1];
            } else if (start2 > end2) {
                return nums1[target-start2];
            }
        } else if (even) {
            if (start1 > end1) {
                return (nums2[target-start1]+nums2[target-start1-1])/2.;
            } else if (start2 > end2) {
                return (nums1[target-start2] + nums1[target-start2-1])/2.;
            } else if (start1 == end1 && start2 == end2) {
                return (nums1[start1] + nums2[start2])/2.;
            }
        }


        // Middle elements (or upper middle elements in even size arrs)
        int m1 = start1+(end1-start1+1)/2;
        int m2 = start2+(end2-start2+1)/2;
        int smallerElms = m1+m2; // Highest index of smaller part elements
        if (even && target == smallerElms+1) { // The problem of one target in lower half and one target in upper half
            if(end1>m1)
                m1++;
            else if(end2>m2)
                m2++;
            else if(start1<m1)
                m1--;
            else if(start2<m2)
                m2--;
            smallerElms = m1+m2;
        }
        

        if(nums1[m1] <= nums2[m2]) {
            if(target <= smallerElms) {
                return findMedianSortedArraysRec(nums1, nums2, start1, start2, end1, m2-1);
            } else {
                return findMedianSortedArraysRec(nums1, nums2, m1+1, start2, end1, end2);
            }
        } else {
            if(target<=smallerElms) {
                return findMedianSortedArraysRec(nums1, nums2, start1, start2, m1-1, end2);
            } else {
                return findMedianSortedArraysRec(nums1, nums2, start1, m2+1, end1, end2);
            }
        }
    }
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        return findMedianSortedArraysRec(nums1, nums2, 0, 0, nums1.size()-1, nums2.size()-1);
    }
};
